File: 0257-binary-tree-paths/0257-binary-tree-paths.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private final List<String> ans = new ArrayList<>();
    private void dfs(TreeNode root, StringBuilder path) {
        if (root == null) return;

        final int tmp = path.length();

        if (root.left == null && root.right == null) {
            path.append(root.val);
            ans.add(path.toString());
        } else {
            path.append(root.val);
            path.append("->");

            dfs(root.left, path);
            dfs(root.right, path);
        }

        path.setLength(tmp);
    }
    public List<String> binaryTreePaths(TreeNode root) {
        dfs(root, new StringBuilder());
        return ans;
    }
}

File: 0257-binary-tree-paths/0257-binary-tree-paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    vector<string> ans;
    void dfs(TreeNode* root, string path) {
        if (!root) return;

        const size_t tmp = path.size();
        path += to_string(root->val);

        if (!root->left && !root->right) {
            ans.push_back(path);
        } else {
            path += "->";
            dfs(root->left, path);
            dfs(root->right, path);
        }

        path.resize(tmp);
    }
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        dfs(root, "");
        return ans;
    }
};

auto init = []() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    cout.tie(nullptr);
    return 'c';
}();